nexus/src/app/certificate.rs

```diff
@@ -47,8 +49,29 @@ impl super::Nexus {
             .silo_required()
             .internal_context("creating a Certificate")?;
 
-        let silo_fq_dns_names =
-            self.silo_fq_dns_names(opctx, authz_silo.id()).await?;
+        // The `opctx` we received is going to be checked for permission to
+        // create a cert below in `db_datastore.certificate_create`, but first
+        // we need to look up this silo's fully-qualified domain names in order
+        // to check that the cert we've been given is valid for this silo.
+        // Looking up DNS names requires reading the DNS configuration of the
+        // _rack_, which this user may not be able to do (even if they have
+        // permission to upload new certs, which almost certainly implies a
+        // silo-level admin. We'll construct a new opctx here to allow this DNS
+        // config lookup, because we believe it does not leak any information
+        // that a silo admin doesn't already know (the external DNS name(s) of
+        // the rack, which leads to their silo's DNS name(s)).
+        //
+        // See https://github.com/oxidecomputer/omicron/issues/4532 for
+        // additional background.
+        let silo_fq_dns_names = {
+            let dns_opctx = OpContext::for_background(
+                opctx.log.clone(),
+                Arc::clone(&self.authz),
+                authn::Context::internal_service_balancer(),
+                Arc::clone(self.datastore()),
+            );
```

Comment:
I see that we've been a little inconsistent on how we create OpContexts that Nexus uses to do privileged things on behalf of unprivileged users.

- For the "instance allocator" and "external authn" OpContexts, we create them once at startup and store them into `Nexus`.  Callers can access the former directly as a field.  They can access the latter that way or using `self.opctx_external_authn()`.
- For the service balancer and internal API OpContexts, we have similar methods `self.opctx_for_service_balancer()` and `self.opctx_for_internal_api()`, but these construct the `OpContext` anew each time rather than just doing that once on startup.

I generally prefer the first approach but clearly it's common to do the latter too.  I was going to suggest _not_ doing this logic inline here, creating a helper instead, although maybe you did this on purpose to preserve the existing request log (in `opctx.log`)?

There's also the question of whether this is the _right_ OpContext.  This definitely _isn't_ the service balancer...but neither is the other user of this OpContext that I found.  In a sense, it really doesn't matter.  These internal identities only exist to limit the privileges of code paths.  You _could_ create a new identity here that can only read DNS and use that.  But it might be overkill to do that everywhere we need to use this pattern.

I see that we have an "internal-read" identity that only gets "fleet viewer", not "fleet admin".  I also see that we already have `Nexus.opctx_alloc` using that identity.  You could use that here and it'd be a little safer (fewer privileges) and also would avoid constructing a new OpContext inline with the request.  The only downside is that the label attached to the log for that one is "InstanceAllocator", which would be a little misleading.  You could also just go ahead and construct a new OpContext here using `authn::Context::internal_read()` instead, or add a helper to do it.

I don't have a strong feeling about all these options.  I would at least use the `internal_read()` identity because there's not much work and I don't think there's any downside.  As for whether to create a new identity, whether to create the OpContext inline or use a helper or use the existing one, these all have tradeoffs and one could go either way.

Reply:
Summarizing an offline chat:

* We can't use "internal-read", because the DNS config policy only grants read to "fleet admin" (which "internal-read" isn't).
* Per [the comments on DnsConfig](https://github.com/oxidecomputer/omicron/blob/07eb7dafc20e35e44edf429fcbb759cbb33edd5f/nexus/db-queries/src/authz/omicron.polar#L355-L359), Nexus uses the "external-authenticator" role when creating silos, so it seems reasonable to use it here as well.

Changed in [fa74a7a](https://github.com/oxidecomputer/omicron/pull/4669/commits/fa74a7aa263a2ed092867b3e952c22300a7e05d5).